### PR TITLE
Show "loading" screen for background reports and display when ready

### LINKED ISSUE #10752: Show generated report to user however long it takes
### What we should change and why (this is tech debt)

After #10732, we display a generic timeout message when a report takes too long and we'll email a link to the report via email when it's done. But we also want to display the report or at least a download link when it's done. We can use cable_ready channel broadcasting for this.

### Context

Follow up from:

* #10732

### PR DESCRIPTION
**Pull request by mkllnk**
#### What? Why?

- Closes #10752 

Originally opened by @Matt-Yorkley against one of my pull requests:

* https://github.com/mkllnk/openfoodnetwork/pull/1

#### What should we test?

- Visit a reports page.
- Compile a report and it should display as usual.
- Activate background reports.
- When you now compile a report, you see a loading screen until the report is ready. It then displays on screen or offers a download link.

#### Release notes

Changelog Category: User facing changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

**Comment by mkllnk** [comment]
@Matt-Yorkley Do you have time to work on this further? There are failing specs and URL generation fails:

> ArgumentError: Cannot generate URL for orders_and_distributors_20230516.pdf using Disk service, please set ActiveStorage::Current.url_options.

**Comment by drummer83** [comment]
Ah, nice! I wasn't aware of this draft!
It may close some of my ideas here: [wishlist #448](https://github.com/openfoodfoundation/wishlist/issues/448)

**Comment by mkllnk** [comment]
Hm, need to work on a flaky spec here.

**Comment by mkllnk** [comment]
> I've used the Sidekiq::Testing.inline! helper with a block around submitting the reports form(s) after noticing @filipefurtad0 mentioning it. It seems like it helps

It means that the report is generated straight away and ready before the controller action ends. So we are not testing the path of a report taking longer... I'll have a look.

**Comment by mkllnk** [comment]
I can reproduce the flaky spec on the previous commit:
```
$ ./script/rspec-slow-repeat 10 spec/system/admin/reports_spec.rb:40
Running 10 times with 4 cores
    !!! Fail !!!
    !!! Fail !!!
Pass.
Pass.
    !!! Fail !!!
Pass.
    !!! Fail !!!
Pass.
    !!! Fail !!!
    !!! Fail !!!
4 of 10 passed (40%)
Exiting...
Check tmp/rspec.log for details.
```

**Comment by mkllnk** [comment]
This spec seems to be stable now but I discovered a race condition. I assume it to be rare in production but a busy server could spark this more often.

I've seen people delaying job executions by a second or so to avoid this but that's bad UX if people always have to wait a second to see the report. There must be a better way. Maybe we can show the loading screen on form submission within the browser instead of via cable car.

**Comment by mkllnk** [comment]
@Matt-Yorkley Executing Sidekiq inline made the problem worse. There's also a bit of contention around its use. And since the flakiness only appears with background reports and cable_car I replaced your two commits with my solution.

**Comment by Matt-Yorkley** [comment]
> I discovered a race condition

Yeah I see what you mean, the controller's response could land after the job finishes assuming the report generation is really fast (although that sounds like an edge case :joy:). 

We could handle it on the client side for sure, or we could change this:
```ruby
def background(format)
  blob = ReportBlob.create_for_upload_later!(report_filename)

  ReportJob.perform_later(
    report_class: report_class, user: spree_current_user, params: params,
    format: format, blob: blob, channel: ScopedChannel.for_id(params[:uuid]),
  )

  render cable_ready: cable_car.
    inner_html(
      selector: "#report-table",
      html: render_to_string(partial: "admin/reports/loading")
    ).scroll_into_view(
      selector: "#report-table",
      block: "start"
    )
end
```

to this...

```ruby
def background(format)
  cable_ready[ScopedChannel.for_id(params[:uuid])]
    .inner_html(
      selector: "#report-table",
      html: render_to_string(partial: "admin/reports/loading")
    ).scroll_into_view(
      selector: "#report-table",
      block: "start"
    ).broadcast

  blob = ReportBlob.create_for_upload_later!(report_filename)

  ReportJob.perform_later(
    report_class: report_class, user: spree_current_user, params: params,
    format: format, blob: blob, channel: ScopedChannel.for_id(params[:uuid]),
  )

  head :no_content
end
```

if you catch my drift...

File: app/webpacker/controllers/scoped_channel_controller.js
Comment (rioug) [inline app/webpacker/controllers/scoped_channel_controller.js:1]: Could we add test for this ? or is it too complicated ?

File: spec/system/admin/reports_spec.rb
Comment (rioug) [inline spec/system/admin/reports_spec.rb:114]: @mkllnk  We could probably mitigate that by having the loading spinner and the report being rendered in two different div? Then the spinner wouldn't override the existing report ? It's not ideal to show a spinner when the report is already there but I think it's a better alternative ? What do you think ?

File: app/webpacker/controllers/scoped_channel_controller.js
Comment (Matt-Yorkley) [reply app/webpacker/controllers/scoped_channel_controller.js:1]: Hmmm that's an interesting question. Looking at it I can't see that there's much that can be reasonably tested here in terms of unit tests. The behaviour is covered by system tests :+1:. Basically the class just connects to an `ActionCable` channel and then doesn't do anything unless the channel tells it to, and it's delegating those bits to other classes anyway... I'm not sure any of that can meaningfully be tested in isolation (outside of a full-stack integration test), or if it could it would all be heavily stubbed/mocked and wouldn't really provide any value..?

**Review by mkllnk — COMMENTED** [review]
@Matt-Yorkley, your patch seems to work. The spec passes 100% on my machine. I can't approve this PR but because it's mainly Matt's work, I think my vote to approve should count. :smile: 

**:heavy_check_mark:**

**Review by abdellani — APPROVED** [review]
looks good to me :)

**Comment by drummer83** [comment]
Hi @mkllnk,
Thanks for the continuing work on this!

I tried the Order Cycle Supplier Totals report as super admin, so with a loooot of data.

With background reports disabled I got an error 500 (timeout) after 120 s.
With background reports enabled I got an (kind of) endless spinner. No notification about an email or anything. I never got to see a link, an email or the report.

While we are working on this, what do you think about this proposal: [wishlist #448](https://github.com/openfoodfoundation/wishlist/issues/448)? I think this would be good usability. What do you think?

I don't see an improvement here, sorry... Am I missing something?

**Comment by mkllnk** [comment]
> I tried the Order Cycle Supplier Totals report as super admin, so with a loooot of data.

You may have used a lot of memory then. When you switch to background reports, it's good to also do a deploy or restart puma so that the memory is freed up again for background reports to use them.

Did you also try a smaller report first? It would be good to know if it works in principle. It did in the past but I'm wondering if any recent merges to master broke something, like the Redis errors we are seeing.

I looked at the log files and the Sidekiq metrics and it seems that 4 background reports were successfully run, taking an average of 2 minutes each.

**Comment by mkllnk** [comment]
Your UX suggestions are great. But I think that they should be handled in a separate pull request.

File: app/webpacker/controllers/scoped_channel_controller.js
Comment (rioug) [reply app/webpacker/controllers/scoped_channel_controller.js:1]: I guess you could test that `CableReady` is called when data is received, but as you point out it would involve stubbing/mocking. 
System tests will pick up any problem with this so that's fine :+1:

**Comment by drummer83** [comment]
Hi @mkllnk,
Thanks for the hint regarding fresh deployment and consumed memory. I will take that into account. :+1: 

Ok, I tried again, this time on staging AU and FR, but we don't have enough data here to run into the timeout (120 s).

So this one is blocked until I can use staging UK again (after merging #10939).

**Comment by drummer83** [comment]
Ok, so after a fresh deploy to staging UK we have the following situation with background reports ON:
| Report generation duration | On screen | Send email notification |
| - | - | - |
| t < 5 seconds | display report :heavy_check_mark: | no :heavy_check_mark: |
| 5 seconds < t < 2 minutes | when report ready: display report :heavy_check_mark: | yes :heavy_check_mark: |
| t > 2 minutes | after 2 min: display email hint (this may fail on slow servers, ignore) :x: | not yet :heavy_check_mark: |
| | when report ready: display report :heavy_check_mark: | yes :heavy_check_mark:  |

I used the reports 'Orders and Distributors' and 'Order Cycle Supplier Totals' and increased the date range.

The 'Orders and Distributors' report with no date limitation at all ended up in an 'endless' spinner and no results were displayed, no email was sent - that is after 20 minutes or so. I will update if it arrives eventually.

## Summary

- In principle the reports are working. 
- Reports taking longer than 2 minutes will be displayed as well (when ready) :heavy_check_mark: 
- The email hint never showed on screen. :x: 
- I may have managed to 'overload' the system (or a worker) with too large data amount - I hope there is a mechanism to recover it automatically. :question: 
- Usability still has room for improvement. :wink: [wishlist #448](https://github.com/openfoodfoundation/wishlist/issues/448)

I will move to 'Ready To Go' but I'm leaving it for you @mkllnk to merge.
Thanks!

**Comment by mkllnk** [comment]
> The email hint never showed on screen.

That's fine. Your wishlist has better ideas for this.

> I may have managed to 'overload' the system (or a worker) with too large data amount

That's still possible but less likely. If this becomes a problem in production, which has a lot more memory, then we may need to invest more work into this. There are ideas already.

**Comment by Matt-Yorkley** [comment]
> I may have managed to 'overload' the system (or a worker) with too large data amount - I hope there is a mechanism to recover it automatically. 

Can you give a bit more detail on this, @drummer83 ? What happened?

**Comment by drummer83** [comment]
> The 'Orders and Distributors' report with no date limitation at all ended up in an 'endless' spinner and no results were displayed, no email was sent - that is after 20 minutes or so. I will update if it arrives eventually.

This is what I experienced as super admin user on staging UK. The email never arrived. @Matt-Yorkley

**Pull request merged**
